lib/puppet/provider/exec/pwsh.rb

```diff
@@ -43,11 +43,28 @@ def validatecmd(command)
 
   private
 
+  # Returns the absolute path to pwsh using a cached result when possible
+  #
+  # @api private
+  # @return [String] the absolute path to the found pwsh executable.  Returns nil when it does not exist
+  def pwsh_command
+    return @pwsh_command if defined? @pwsh_command
```

Comment:
Unfortunately I can't use that because `nil` is a valid value (indicates pwsh could not be found)

Reply:
Taking a step back and given the only caller will immediately call `self.fail if pwsh.nil?` I'm not sure why this intermediate `pwsh_command` method is even necessary? i.e. in `run` I think you can just do this instead.


```
def run(command, check = false)
  # will reuse cached value on subsequent invocation of this resource instance
  @pwsh ||= get_pwsh_command

  # because it will fail here anyhow if the value is nil
  self.fail 'pwsh could not be found' if @pwsh.nil?

  ...
```

The other thing is that this is not a global caching anyhow (which is where you'd get bigger savings), but rather per provider instance (i.e. one time per resource) since we have to vary things dynamically based on presence of `:path`.... In other words this just makes sure that when a `command` has an `unless` or `onlyif` that we only find the command once.

Reply:
Yeah true, and the caching only _really_ helps when pwsh is missing (Doesn't search multiple) times.  Once it's found, using the `||=` short circuits that.

So, ok I'll remove that caching bit.  And if it truly is a problem we'll need to memoize globally.

Reply:
Done.  Thanks again for taking to time to review.